### ISSUE TITLE
Support AWS Lambda Aliases and Versions

### ISSUE THREAD

**Issue by johannesboyne** [hashicorp/terraform#4123]
I would like to be able to use `terraform` to configure the aliases and versions for Lambda functions.

- Create Alias: https://github.com/aws/aws-sdk-go/blob/master/service/lambda/api.go#L59
- Publish Version: https://github.com/aws/aws-sdk-go/blob/master/service/lambda/api.go#L662

I'll try to squeeze in some time to start working on it, but as the terraform acc. test quality is quite high - kudos ;-) - I'll first have to read and understand parts of the testing code.

**Comment by johannesboyne** [comment]
This is the first time I'm contributing to terraform and thus I'm not 100% sure whether everything is done best.
In commit https://github.com/johannesboyne/terraform/commit/0fe06259d7de7dad672a52dd8c935fb1ed4875e5  the lambda function alias support is implemented and I'll continue with function versions later. Just, if somebody would likes to comment on my attempt, please feel free to do so!
When I'm done I would squash aliases and versions into one single commit, if you don't ask me to separate it.

**Comment by johannesboyne** [comment]
I would like to get some feedback on the usage of versions. I cam up with the following:
```tf
resource "aws_lambda_function" "lambda_function_test" {
    filename = "test-fixtures/lambdatest.zip"
    function_name = "example_lambda_version_test_name"
    role = "${aws_iam_role.iam_for_lambda.arn}"
    handler = "exports.example"
}

resource "aws_lambda_version" "current_version" {
    function_name = "${aws_lambda_function.lambda_function_test.arn}"
    description = "My description for this version"
}
```

A version would be pushed with the function creation / update and one could later access it e.g. in an API Gateway configuration. **BUT** and this _but_ bothers me quite a bit: would anyone actually use it programmatically like this? The only _"reasonable"_ use-case is rolling back to a version because of errors in the Lambda function. But I guess most of the time `s3_object_version` makes more sense for something like that.

Any opinions on the lambda versioning?

**Comment by jen20** [comment]
Hi @johannesboyne! Thanks for opening up an issue, that's definitely the best route to starting to contribute to Terraform. I had a look through your code, and with the exception of the things you commented on yourself, it looks like a pretty good start to me. If you'd like to open it as a pull request and mention this issue number, we can review it properly!

Although the need for setting up aliases for the event source mapping is somewhat reduced by using Terraform to manage them (I think), the recommendation [from Amazon](http://docs.aws.amazon.com/lambda/latest/dg/versioning-aliases.html) is that aliases should be used. Consequently I think this is a good addition to Terraform's coverage of the AWS API.

**Comment by johannesboyne** [comment]
Hi @jen20 great, I'll do that. 

My main reason for aliases has been the combination with _api gateway_.

**Comment by grisha** [comment]
I'm a Terraform noob, so it's entirely possible that I'm missing something obvious, but as far as I can tell, there is also no mechanism by which the Lambda zip file would be deployed if the zip file has changed, the `resourceAwsLambdaFunctionUpdate()` looks like it's not yet implemented: https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/resource_aws_lambda_function.go#L235-L239. It would also be nice if some method existed by which the zip file could be built locally, e.g. if the Lambda contains code that needs to be generated (or a component of the zip file compiled, e.g. a Go program) before apply time.

**Comment by mostman79** [comment]
@johannesboyne - this is really good, we will use this for sure once it becomes integrated. Thanks for taking this on. To answer your question on if/when you would use that, Yes :)

Being able to call older versions in places where you are programmatically calling functions (vs trigger or API), is very important to tracking down regressions.

@grisha - agreed that all of those things would be great, here is what I did to work around that missing functionality. I wrote a set of tasks (grunt tasks in my case) that did the following:

1. Zip up the files. For me, this was non trivial, I have multiple Lambda functions in a single repo, so keeping it all straight was a pain. The zips end up in a build/ folder.
2. Run an md5 function on each zip and slap the md5 output on the filename.
3. Rebuild your terraform config so that the filename field of the lambda function definition reflects the newly generated zip.
4. Apply 

This will result in a setup where your lamdba functions are always (and only) updated if their code changes. Not ideal, but it works.
